File: cases/taint/ruby/interprocedural-exception-persistence-positive/interprocedural_exception_persistence_positive.rb
def dfb_source # DFB-SOURCE: exception-persistence-input
  "tainted"
end

def dfb_sink(value) # DFB-SINK: exception-persistence-sink
end

class FlowBox
  attr_accessor :value

  def initialize(value)
    @value = value
  end
end

class FlowError < StandardError
end

def store_and_throw(box, value)
  box.value = value # DFB-WITNESS: exception-persistence-store
  raise FlowError, "exceptional exit" # DFB-WITNESS: exception-persistence-throw
end

def recover(box, value)
  begin
    store_and_throw(box, value)
    "unreachable"
  rescue FlowError
    box.value # DFB-WITNESS: exception-persistence-recovery
  end
end

def run
  dfb_sink(recover(FlowBox.new("seed"), dfb_source))
end

File: cases/taint/javascript/interprocedural-exception-persistence-positive/interprocedural_exception_persistence_positive.js
function dfb_source() { // DFB-SOURCE: exception-persistence-input
  return "tainted";
}

function dfb_sink(value) {} // DFB-SINK: exception-persistence-sink

class FlowError extends Error {}

class FlowBox {
  constructor(value) {
    this.value = value;
  }
}

function store_and_throw(box, value) {
  box.value = value; // DFB-WITNESS: exception-persistence-store
  throw new FlowError("exceptional exit"); // DFB-WITNESS: exception-persistence-throw
}

function recover(box, value) {
  try {
    store_and_throw(box, value);
    return "unreachable";
  } catch (error) {
    if (!(error instanceof FlowError)) throw error;
    return box.value; // DFB-WITNESS: exception-persistence-recovery
  }
}

function run() {
  dfb_sink(recover(new FlowBox("seed"), dfb_source()));
}
